README
Unicode/Japanese version 0.20
=============================

Unicode::Japanese supports inter-convertion between sjis(CP932),
euc-jp, jis, unicode(utf8,etc) and "EMOJI" of the mobile phones.

DoCoMo i-mode, ASTEL dot-i, Vodafone Vodafone-live!(was J-PHONE
J-Sky).


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

If you have no C-Compiler or no need XS modules:

   perl Makefile.PL PurePerl
   make
   make test
   make install

DEPENDENCIES

no dependencies, exclude default modules.

COPYRIGHT AND LICENCE

Copyright 2001-2004
SANO Taku (SAWATARI Mikage) and YAMASHINA Hio.
All right reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for Perl extension Unicode::Japanese.

0.46 Fri Aug 29 16:55:29 JST 2008
    - libunijp, erlang-support.
    - fix, static methods are failed until creattion of an instance.

0.45 Mon Jan 21 17:32:36 JST 2008
    - Fixed enormous english error messages and documentation which were
      grammatically incorrect or way too hard to be understood.
    - fix, old perl compatibility (now available on perl-5.004 and later).
    - add, utf8-imode encoding.
    - fix, null chars were show when input is invalid sjis-imode.
    - fix, guess ordering on xs (imode-doti-au => au-imode-doti to comform to doing on pp)
    - fix, some doti text was guessed as au text on pp mode when it begin with au text.

0.44 Fri Sep 14 14:26:22 JST 2007
	- fix, broken emoji convert on alpha cpu.
	- fix, t/pod.t failed before perl 5.8.0.

0.43 Mon Sep 10 16:17:09 JST 2007
	- add, jsky encoding with no ending escape.

0.42 Mon Sep  3 16:39:31 JST 2007
	- fix, broken sjis-au convert in some case.

0.41 Fri Aug 31 18:11:40 JST 2007
	- add sjis-au, sjis-au1, sjis-au2 encodings.
	- add utf8-jsky, utf8-jsky1, utf8-jsky2 encodings.

0.40_01
        - fix, broken on perl-5.6.1.

0.40 Tue Jan 16 09:54:13 JST 2007
        - fix, pod.

0.39 Fri Jan 12 18:33:54 JST 2007
        - fix, getcode could not detect imode-pictgrams extend 61st-63rd.

0.38 Tue Sep  5 14:53:19 JST 2006
        - fix, netinet/in.h not required.

0.37 Mon Jul  3 10:38:01 JST 2006
        - fix, sparc sigbus. thx kohju ;)

0.36 Tue Jun 20 19:09:43 JST 2006
        - fix, japanese document was shown on cpan.

0.35 Mon Jun 19 17:47:54 JST 2006
        - made validate_utf8 available in XS.
        - add, -l switch for ujconv(1).
        - fix, au mapping broken.

0.34 Thu Mar 23 17:16:32 JST 2006
	- fix, JIS-X6226-1979 escape sequence (\e$@) on xsubs.

0.33 Sat Feb  4 15:51:37 JST 2006
	- fix,  typos in pod.
	- add,  unijp, a short-named constructor.

0.32 Fri Nov  4 12:01:37 JST 2005
	- fix, some au/vodafone emoji mapping failure.

0.31 Wed Sep 28 22:15:42 JST 2005
	- fix, getcode of jis-kana on xs was broken.

0.30 Mon Sep  5 12:53:51 JST 2005
	- add, au emoji.
	- rewrite mmap code on XSubs.
	- reduce mem/cpu on huge string and most of that is ascii.

0.27 Sun May 15 18:57:51 JST 2005
	- support for x86_64 (EM64T)
	- fix, surrogate pair was not worked

0.26 Tue Feb 15 13:16:30 JST 2005
	- fix, 0.25 was not updated on xs data.
	- add, cp932([fa-fc]xx) map into jis.

0.25 Tue Feb 15 13:00:16 JST 2005
        - bin/ujconv and bin/ujguess added.
	- unicode char converted into smaller CP932 code.

0.24 Mon Jan 24 11:23:05 JST 2005
        - avoid unexpected upgrade to imode2/jsky2 from imode1/jsky1.

0.23 Thu Nov  4 16:16:48 JST 2004
        - avoid illegal utf-8 sequence
        - add mapping, U+301C (WAVE TILDE) into CP932:8160

0.22 Mon May 31 18:48:42 JST 2004
        - Bug Fix, on 5.8.4, join_csv/split_csv raise SEGV.

0.21 Tue May 25 23:59:15 JST 2004
        - Bug Fix, jis.roman escape sequence is not processed on XS code.

0.20 Sun Mar  7 19:08:55 JST 2004
        - Bug Fix, when pure-perl, make test had failed.
	- untaint.

0.19 Fri Jan 16 19:48:46 JST 2004
	- Bug Fix, decoding empty string from jis becomes "\0".

0.18 Wed Nov  6 18:54:32 JST 2002
        - Bug Fix, Bus Error on sparc.

0.17 Thu Oct 31 20:46:45 JST 2002
	- Bug Fix, compile failure on perl-5.6.0 prior.
        - support MSWin32 XS.
	
0.16 Tue Oct 29 16:25:11 JST 2002
        - excludes c++ code, now XS is written only c.
        
0.15  Mon Oct 28 00:00:32 JST 2002
        - new emoji characters. imode2(imode extended chars), jsky2(jsky
          page 3 to 6).
        - fix warning at used in -w environment.
        - new method, getu().

0.14  Sat Aug 31 15:30:20 JST 2002
        - Supports in `use utf8' scope.
        - Fix h2z's miss-conversion for "'".

0.13  Wed Jul 24 01:26:53 JST 2002
	- Fix problem which fails in compile in some environments.

0.12  Mon Jul 15 15:29:16 JST 2002
	- Fix problem which fails in compile in some environments.

0.10  Tue Jul  9 18:00:52 JST 2002
	- Fix creates duplicate copies of blib/lib under blib/script,
	  all of which are owned by root.

0.09  Mon Jul  8 16:54:01 JST 2002
	- Many fixes (included vulnerability of DoS)
	- Supports imode's extended EMOJI (they only can be mapped to
	  Unicode area. Convertion among each model is not supported yet)
	- New XS mode implementation; convertion utf8<->ucs2.
	- Can specify PurePerl mode when loading with 'PurePrel' keyword.

0.08  Tue Jun 18 13:45:18 JST 2002
	- Supports XS. 
	- Some bugs has been fixed.

0.03  Mon Sep 17 17:16:01 JST 2001
	- t/getcode.t added.
	- Manual updated. (It translated into English.)
	- Fixed bug, doesn't work with perl 5.005. 
	- Fixed bug, $_ is broken when first method call.

0.02  Wed Aug  1 17:48:33 JST 2001
	- Manual updated.
	- Changed the handling of characters which out of SJIS codes' range
	  in converting from UTF8 to SJIS.
	- The time of loading the module was reduced. (with AUTOLOAD function)
	- t/outrange.t added.

0.01  Fri Jul 27 20:51:44 JST 2001
	- BETA Release.



ucs4.c
/* ----------------------------------------------------------------------------
 * ucs4.c
 * ----------------------------------------------------------------------------
 * Mastering programed by YAMASHINA Hio
 * ----------------------------------------------------------------------------
 * $Id: ucs4.c 41492 2008-02-15 08:26:18Z hio $
 * ------------------------------------------------------------------------- */


#include "Japanese.h"

/* ----------------------------------------------------------------------------
 * convert ucs4 into utf-8
 * ------------------------------------------------------------------------- */
EXTERN_C
SV*
xs_ucs4_utf8(SV* sv_str)
{
  UJ_UINT8* src;
  int len;
  SV_Buf result;
  const UJ_UINT8* src_end;
  UJ_UINT8 buf[4];

  if( sv_str==&PL_sv_undef )
  {
    return newSVpvn("",0);
  }
  
  src = (UJ_UINT8*)SvPV(sv_str,PL_na);
  len = sv_len(sv_str);
  src_end = src+(len&~1);
  /*fprintf(stderr,"Unicode::Japanese::(xs)ucs4_utf8\n",len);*/
  /*bin_dump("in ",src,len);*/
  SV_Buf_init(&result,len*3/2+4);

  if( len&3 )
  {
    Perl_croak(aTHX_ "Unicode::Japanese::ucs4_utf8, invalid length (not 4*n)");
  }

  for(; src<src_end; src+=4 )
  {
    const UJ_UINT32 ucs4 = (src[0]<<24)+(src[1]<<16)+(src[2]<<8)+src[3]; /* ntohs */
    if( ucs4<0x80 )
    {
      SV_Buf_append_ch(&result,(UJ_UINT8)ucs4);
    }else if( ucs4<0x800 )
    {
      buf[0] = 0xC0 | (ucs4 >> 6);
      buf[1] = 0x80 | (ucs4 & 0x3F);
      SV_Buf_append_mem(&result, buf, 2);
    }else if( ucs4 < 0x10000 )
    {
      buf[0] = 0xE0 | (ucs4 >> 12);
      buf[1] = 0x80 | ((ucs4 >> 6) & 0x3F);
      buf[2] = 0x80 | (ucs4 & 0x3F);
      SV_Buf_append_mem(&result, buf, 3);
    }else if( ucs4 <= 0x0010FFFF )
    {
      buf[0] = 0xF0 |  (ucs4 >> 18);
      buf[1] = 0x80 | ((ucs4 >> 12) & 0x3F);
      buf[2] = 0x80 | ((ucs4 >>  6) & 0x3F);
      buf[3] = 0x80 | ( ucs4        & 0x3F);
      SV_Buf_append_mem(&result, buf, 4);
    }else
    {
      SV_Buf_append_ch(&result,'?');
    }
  }

  /*bin_dump("out",result.getBegin(),result.getLength()); */
  SV_Buf_setLength(&result);

  return SV_Buf_getSv(&result);
}

/* ----------------------------------------------------------------------------
 * convert utf-8 into ucs4
 * ------------------------------------------------------------------------- */
EXTERN_C
SV*
xs_utf8_ucs4(SV* sv_str)
{
  UJ_UINT8* src;
  int len;
  SV_Buf result;
  const UJ_UINT8* src_end;
  const UJ_UINT8 buf_failed[4] = { 0, 0, 0, '?' };

  if( sv_str==&PL_sv_undef )
  {
    return newSVpvn("",0);
  }
  
  src = (UJ_UINT8*)SvPV(sv_str,PL_na);
  len = sv_len(sv_str);
  src_end = src+len;
  /*fprintf(stderr,"Unicode::Japanese::(xs)utf8_ucs4\n",len); */
  /*bin_dump("in ",src,len); */
  SV_Buf_init(&result,len*4);
  
  while( src<src_end )
  {
    UJ_UINT32 ucs;
    if( *src<=0x7f )
    { /* ascii. */
      SV_Buf_append_ch4(&result,htonl(*src));
      ++src;
      continue;
    }
    if( 0xc0<=*src && *src<=0xdf )
    { /* length [2] */
      const int       utf8_len = 2;
      const UJ_UINT32 ucs_min  = 0x80;
      const UJ_UINT32 ucs_max  = 0x7ff;
      if( src+1>=src_end ||
          src[1]<0x80 || 0xbf<src[1] )
      {
        SV_Buf_append_ch4(&result,htonl(*src));
        ++src;
        continue;
      }
      
      /* compute code point */
      ucs = ((src[0] & 0x1F)<<6)|(src[1] & 0x3F);
      src += utf8_len;
      if( ucs_min<=ucs && ucs<=ucs_max )
      { /* noop */
      }else
      { /* illegal sequence */
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        continue;
      }
      
      /* ok. */
    }else if( 0xe0<=*src && *src<=0xef )
    { /* length [3] */
      const int       utf8_len = 3;
      const UJ_UINT32 ucs_min  = 0x800;
      const UJ_UINT32 ucs_max  = 0xffff;
      if( src+2>=src_end ||
          src[1]<0x80 || 0xbf<src[1] ||
          src[2]<0x80 || 0xbf<src[2] )
      {
        SV_Buf_append_ch4(&result,htonl(*src));
        ++src;
        continue;
      }
      
      /* compute code point */
      ucs = ((src[0] & 0x0F)<<12)|((src[1] & 0x3F)<<6)|(src[2] & 0x3F);
      src += utf8_len;
      if( ucs_min<=ucs && ucs<=ucs_max )
      { /* noop */
      }else
      { /* illegal sequence */
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        continue;
      }
      
      if( ucs<0xD800 || ucs>0xDBFF )
      { /* normal char, noop */
      }else
      { /* delete surrogate pair range */
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        continue;
      }
      
      /* ok. */
    }else if( 0xf0<=*src && *src<=0xf7 )
    { /* length [4] */
      const int       utf8_len = 4;
      const UJ_UINT32 ucs_min  = 0x010000;
      const UJ_UINT32 ucs_max  = 0x10ffff;
      if( src+3>=src_end ||
          src[1]<0x80 || 0xbf<src[1] ||
          src[2]<0x80 || 0xbf<src[2] ||
          src[3]<0x80 || 0xbf<src[3] )
      {
        SV_Buf_append_ch4(&result,htonl(*src));
        ++src;
        continue;
      }
      
      /* compute code point */
      ucs = ((src[0] & 0x07)<<18)|((src[1] & 0x3F)<<12)|
             ((src[2] & 0x3f) << 6)|(src[3] & 0x3F);
      src += utf8_len;
      if( ucs_min<=ucs && ucs<=ucs_max )
      { /* noop */
      }else
      { /* illegal sequence */
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        continue;
      }
      
      /* ok. */
    }else if( 0xf8<=*src && *src<=0xfb )
    {
      const int          utf8_len = 5;
      if( src+utf8_len<=src_end )
      { /* noop */
      }else
      { /* no enough sequence */
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        ++src;
        continue;
      }
      /* check follow sequences */
      if( 0x80<=src[1] && src[1]<=0xbf && 0x80<=src[2] && src[2]<=0xbf
          && 0x80<=src[3] && src[3]<=0xbf && 0x80<=src[4] && src[4]<=0xbf )
      { /* noop */
      }else
      {
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        ++src;
        continue;
      }
      
      /* compute code point */
      src += utf8_len;
      SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
      continue;
    }else if( 0xfc<=*src && *src<=0xfd )
    {
      const int          utf8_len = 6;
      if( src+utf8_len<=src_end )
      { /* noop */
      }else
      { /* no enough sequence */
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        ++src;
        continue;
      }
      /* check follow sequences */
      if( 0x80<=src[1] && src[1]<=0xbf && 0x80<=src[2] && src[2]<=0xbf
          && 0x80<=src[3] && src[3]<=0xbf && 0x80<=src[4] && src[4]<=0xbf
          && 0x80<=src[5] && src[5]<=0xbf )
      { /* noop */
      }else
      {
        SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
        ++src;
        continue;
      }
      
      /* compute code point */
      src += utf8_len;
      SV_Buf_append_ch4(&result, *(UJ_UINT32*)buf_failed);
      continue;
    }else
    { /* invalid */
      SV_Buf_append_ch4(&result,htonl(*src));
      ++src;
      continue;
    }

    SV_Buf_append_ch4(&result,htonl(ucs));
    /*bin_dump("now",dst_begin,dst-dst_begin); */
  }

  /*bin_dump("out",result.getBegin(),result.getLength()); */
  SV_Buf_setLength(&result);

  return SV_Buf_getSv(&result);
}

/* ----------------------------------------------------------------------------
 * End Of File.
 * ------------------------------------------------------------------------- */
